### DISCUSSION TITLE
Keycloak behind reverse proxy incredibly slow

### DISCUSSION THREAD

**Discussion by ASullivan219** [keycloak/keycloak#34677] [Q&A]
Hello all, I am having some issues running keycloak behind an nginx reverse proxy when hosted on AWS.

The network flow happens like this:
- client -> AWS network load balancer -> nginx (listening port 443) -> tls -> keycloak

The nginx container also serves the built files for our react front-end. Both nginx and keycloak are being deployed to AWS fargate containers.

I am getting extremely slow response times when authenticating to my realm and when I access the keycloak admin panel. Some requests taking around a minute to resolve. Here is a log from nginx:

```
remoteAddr: [<IP addr removed>] remoteUser: [-] timeLocal: [05/Nov/2024:21:42:20 +0000] request: [GET /kc/realms/iros/.well-known/openid-configuration HTTP/2.0] status: [200] bytesSent: [6831] req_time: [60.325] ,upstream_connect_time: [-, 0.035], upstream_resp_time: [60.002, 0.324] ,upstream_header_time: [-, 0.324]
```

Notice the upstream response time of 60s. I am assuming that this latency must be due to mis configured reverse proxy or perhaps a mis-configured Keycloak set-up.

Below is a sample from my nginx config, some configuration options may seem strange due to the fact that I have been trying to increase performance, but am not well versed in nginx configuration options.

```
        listen 443 ssl http2;
        ssl_certificate /etc/ssl/certs/localhost.crt;
        ssl_certificate_key /etc/ssl/private/localhost.key;
        ssl_verify_client optional_no_ca;
        ssl_session_cache   shared:SSL:10m;
        ssl_session_timeout 10m;
        proxy_ssl_verify off; 
        proxy_buffering off;

        access_log /var/log/nginx/access.log custom;

        location /kc {

            proxy_set_header Host                   $http_host;
            proxy_set_header X-Real-IP              $remote_addr;
            proxy_set_header X-forwarded-Host       $http_host;
            proxy_set_header X-Forwarded-For        $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto      $scheme;
            proxy_set_header ssl-client-cert        $ssl_client_escaped_cert;
            proxy_redirect off;

            proxy_pass https://keycloak.:8443;
        }
```

Below are some Environment variables/ run time flags for the keycloak deployment. For reference I am currently on v25.0.0

```
"KC_HOSTNAME=https://localhost:8000/kc",
     "KC_HTTPS_CERTIFICATE_FILE=/opt/keycloak/conf/certs/certfile.pem",
    "KC_HTTPS_CERTIFICATE_KEY_FILE=/opt/keycloak/conf/certs/keyfile.pem",
    "KC_HEALTH_ENABLED=true",
    "KC_METRICS_ENABLED=true",
    "KC_HOSTNAME_STRICT=false",
    "KC_HTTP_RELATIVE_PATH=/kc",
    "KC_PROXY_PROTOCOL_ENABLED=true",
    "KC_PROXY_HEADERS=xforwarded",
    "KC_TLS_HOSTNAME_VERIFIER=ANY",
    "KC_HTTPS_CLIENT_AUTH=request"

"start",
    "--verbose",
    "--import-realm",
    "--spi-x509cert-lookup-provider=nginx",
    "--spi-x509cert-lookup-nginx-ssl-client-cert=ssl-client-cert",
    "--spi-x509cert-lookup-nginx-certificate-chain-length=2"
```

These options work fine on our local deployment, but when deployed to AWS Fargate the response times essentially render the website unusable. Please let me know if anyone has any tips!

Thank you!

**Comment by embesozzi** [comment]
In relation to NGINX, if you are running it as an Ingress, it should be fine the default config for testing proposes. If you are running it as a container, you should pay attention to the global configuration, such as `worker_processes`, `worker_connections`, `proxy_buffer_size`, etc

**Reply by ASullivan219** [reply]
I appreciate the response on this and will definitely look into global nginx config. If you have any good resources on this topic I would appreciate you sending them my way!

I was able to alleviate some the performance issues by moving the keycloak container into the same aws subnet as the nginx container. Im not sure why this helped, perhaps some type of mis-configuration with our VPC perhaps.

**Reply by embesozzi** [reply]
Here are some examples:
- https://blog.nginx.org/blog/performance-tuning-tips-tricks
- https://gist.github.com/v0lkan/90fcb83c86918732b894

**Comment by divyanshuss** [comment]
Hello @ASullivan219 , 

Thanks for raising this. I am encountering a similar problem. We are using a load balancer ---> HaProxy ----> GCP Virtual Machine. 
Keycloak works well when deployed locally using docker. 

However, this setup is extremely slow. We have deployed it over bare metal. The database is hosted on the same server. 

Keycloak Version = 26.0.0

Do we have any improvements? 

Thank you.
